library/patient.inc.php

```diff
@@ -1820,21 +1824,25 @@ function is_patient_deceased($pid, $date = '')
     }
 }
 
-// This computes, sets and returns the dup score for the given patient.
+// This computes, sets and returns the dup score for the given patient, compared to all patients earlier in the patient_data table
+// for updateing after patients demographics have been changed need to check against whole table, i.e. patients who come earlier and later in the table, i.e. where p2 != p1, this works for new patients as well
 //
 function updateDupScore($pid)
 {
+        // new patient or patient with editted demographics
     $row = sqlQuery(
         "SELECT MAX(" . getDupScoreSQL() . ") AS dupscore " .
         "FROM patient_data AS p1, patient_data AS p2 WHERE " .
-        "p1.pid = ? AND p2.pid < p1.pid",
```

Comment:
@bradymiller This makes sense to check the whole table.  Was there a reason you wrote the code to only check against earlier pids.

Reply:
I just migrated this whole script from patient.inc to patient.inc.php. Looks like @sunsetsystems created this function:
https://github.com/openemr/openemr/blame/rel-700/library/patient.inc

this PR looks good to me if testing well

Reply:
Dupscore was defined/designed to be a comparison only with earlier pids. If you don't like that definition then you need to recompute scores for all pids on every demographics update, and that would be computationally prohibitive. More practical, I think, to run a nightly or weekly process to recompute all scores.

Reply:
@sunsetsystems hi rod. when a patient's demographic are changed their duplicate score is recalculated looking at all the other patients, but this is done only for the patient whose demographics were changed. I don't see a need to do it for every patient in the system at this point.

would it help to have two buttons on editing a patient's demographics: one 'save' and one 'save and update duplicate score' (or something) ? (obviously the save wouldn't do any recalculations). 

if this is useful @adunsulag and @sunsetsystems  then let me know and i'll implement it.

Reply:
Hi Ruth. As I noted you can't recompute everything on each demographics update. In a somewhat large database that can take hours. This is why I suggest a nightly or weekly process. Also for comparing every patient with every other patient it's sufficient for such a process to compare only with lesser pids, which cuts the time in half. You can do what you originally suggested but I think it's better to have a more complete solution where a full recomputation is done before a manual session of managing duplicates. I hope that makes sense.

Reply:
hi rod, i think we are still talking at cross purposes. what we find works is that when a patient's info is changed to compare just that one patient with all the others, in order to recompute their dup score. comparing to only those with lower pids has proved not to work: If the change of a patient makes them a dup of a patient with a higher pid this won't show, however comparing that one patient to all in the data base works in our situation, with no need to recompare every patient with every other (which i think is your suggestion). This is why i suggested the buttons, because then one can use the technique that suites one's own configuration. 

interestingly, what i suggest above is no more compute heavy than when a new patient is added to the system - when an automatic comparison with the whole database is carried out in order to compute their initial dup score. 

perhaps we also need a 'compare all' button/function somewhere which does what you suggest, i.e. recompute the dup score for every patient in the system, which as you rightly say, could take a long time depending on the size of the db. If people are editing patient's data without recomputing ('save' button)  that patient's dup score, then this 'compare all' function would be needed - or some way to run it via cron, as an example.

Reply:
Hi Ruth, the problem with your solution is that it will potentially created "orphaned" instances of high score values that the GUI tool will not show the matches for. It may help you in some cases but it's not robust.

See https://github.com/openemr/openemr/pull/4502 for the original PR and details of the design. The correct way to solve your problem is by periodically running dupscore.cli.php which is also noted in that PR. Yes it works with cron.

Reply:
oh that's interesting, because one of the bugs we found was exactly that - orphaned instances of high score values that were turning up in the gui tool (manage duplicates). so far my new code has not thrown up any, but we are still testing. I'll take a look at issue #4502 - thanks for the lead.

Reply:
As I recall when a patient is marked unique, that is with respect to previously created patients (lower pids). They are still fair game to be compared with more recently created patients.

But more importantly, testing during production is not enough to assure that something does or doesn't work. Furthermore you must run dupscore.cli.php initially, and periodically after that to ensure accuracy.